// gigachat/api/get_image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get_image = get_image;
const utils_1 = require("./utils.js");
const exceptions_1 = require("../exceptions.js");
function getRequestConfig({ fileId, accessToken }) {
    const headers = (0, utils_1.buildHeaders)(accessToken);
    headers['Accept'] = 'application/jpg';
    return {
        method: 'GET',
        url: `/files/${fileId}/content`,
        headers: headers,
        responseEncoding: 'binary',
    };
}
function buildResponse(response) {
    if (response.status === 200) {
        return (0, utils_1.buildXHeaders)(response, { content: response.data });
    }
    else if (response.status === 401) {
        throw new exceptions_1.AuthenticationError(response);
    }
    else {
        throw new exceptions_1.ResponseError(response);
    }
}
async function get_image(client, args) {
    const config = getRequestConfig(args);
    const response = await client.request(config);
    return buildResponse(response);
}
