// gigachat/api/post_tokens_count.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.post_tokens_count = post_tokens_count;
const utils_1 = require("./utils.js");
const exceptions_1 = require("../exceptions.js");
function getRequestConfig({ input, model, accessToken }) {
    const headers = (0, utils_1.buildHeaders)(accessToken);
    const data = { model, input };
    return {
        method: 'POST',
        url: '/tokens/count',
        headers: headers,
        data: data,
    };
}
function buildResponse(response) {
    if (response.status === 200) {
        return (0, utils_1.buildXHeaders)(response, { tokens: response.data });
    }
    else if (response.status === 401) {
        throw new exceptions_1.AuthenticationError(response);
    }
    else {
        throw new exceptions_1.ResponseError(response);
    }
}
async function post_tokens_count(client, args) {
    const config = getRequestConfig(args);
    const response = await client.request(config);
    return buildResponse(response);
}
