// gigachat/api/post_files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.post_files = post_files;
const utils_1 = require("./utils.js");
const exceptions_1 = require("../exceptions.js");
function getRequestConfig({ file, purpose = 'general', accessToken }) {
    const headers = (0, utils_1.buildHeaders)(accessToken);
    const formData = new FormData();
    formData.append('file', file);
    formData.append('purpose', purpose);
    return {
        method: 'POST',
        url: '/files',
        data: formData,
        headers: headers,
    };
}
function buildResponse(response) {
    if (response.status === 200) {
        return (0, utils_1.buildXHeaders)(response, response.data);
    }
    else if (response.status === 401) {
        throw new exceptions_1.AuthenticationError(response);
    }
    else {
        throw new exceptions_1.ResponseError(response);
    }
}
async function post_files(client, args) {
    const config = getRequestConfig(args);
    const response = await client.request(config);
    return buildResponse(response);
}
