// gigachat/api/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.USER_AGENT = void 0;
exports.buildHeaders = buildHeaders;
exports.parseChunk = parseChunk;
exports.buildXHeaders = buildXHeaders;
exports.isRunningInBrowser = isRunningInBrowser;
const constants_1 = require("../constants.js");
Object.defineProperty(exports, "USER_AGENT", { enumerable: true, get: function () { return constants_1.USER_AGENT; } });
function buildHeaders(accessToken) {
    const headers = {};
    if (accessToken) {
        headers['Authorization'] = `Bearer ${accessToken}`;
    }
    if (!isRunningInBrowser()) {
        headers['User-Agent'] = constants_1.USER_AGENT;
    }
    else {
        headers['X-User-Agent'] = constants_1.USER_AGENT;
    }
    return headers;
}
function parseChunk(line) {
    try {
        const idx = line.indexOf(': ');
        if (idx === -1)
            return null;
        const name = line.substring(0, idx);
        const value = line.substring(idx + 2);
        if (name === 'data') {
            if (value === '[DONE]') {
                return null;
            }
            else {
                return value ? JSON.parse(value) : null;
            }
        }
    }
    catch (e) {
        console.error(`Error parsing chunk from server: ${e}, raw value: ${line}`);
        throw e;
    }
    return null;
}
function buildXHeaders(response, data) {
    const headers = response.headers;
    const withXHeaders = {
        xHeaders: {},
    };
    if (headers) {
        withXHeaders.xHeaders = {
            xRequestID: headers['x-request-id'],
            xSessionID: headers['x-session-id'],
            xClientID: headers['x-client-id'],
        };
    }
    return {
        ...data,
        ...withXHeaders,
    };
}
function isRunningInBrowser() {
    return (
    // @ts-ignore
    typeof window !== 'undefined' &&
        // @ts-ignore
        typeof window.document !== 'undefined' &&
        // @ts-ignore
        typeof navigator !== 'undefined');
}
